wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java

```diff
@@ -171,36 +195,70 @@ private Connection connectInternal(
           new Object[] {selectedHostSpec.getHost()}));
       selectedHostSpec.setAvailability(HostAvailability.NOT_AVAILABLE);
       // Retry connect prioritising healthiest router for best chance of connection over load-balancing with round-robin
-      return retryConnectWithLeastLoadedRouters(limitlessRouters, props, e);
+      return retryConnectWithLeastLoadedRouters(limitlessRouters, props, conn, hostSpec, e);
     }
   }
 
+  private Connection connectInternalWithoutDialect(
+      final @NonNull String driverProtocol,
+      final @NonNull HostSpec hostSpec,
+      final @NonNull Properties props,
+      final boolean isInitialConnection,
+      final JdbcCallable<Connection, SQLException> connectFunc) throws SQLException {
+
+    final Connection conn = connectFunc.call();
+
+    final Dialect dialect = this.pluginService.getDialect();
+    if (!(dialect instanceof AuroraLimitlessDialect)) {
+      throw new UnsupportedOperationException(Messages.get("LimitlessConnectionPlugin.unsupportedDialectOrDatabase",
+          new Object[] {dialect}));
+    }
+
+    initLimitlessRouterMonitorService();
+    if (isInitialConnection) {
+      this.limitlessRouterService
+          .startMonitoring(hostSpec, properties, INTERVAL_MILLIS.getInteger(properties));
+    }
+
+    List<HostSpec> limitlessRouters = this.limitlessRouterService.getLimitlessRouters(
+        this.pluginService.getHostListProvider().getClusterId(), props);
+    if (limitlessRouters.isEmpty()) {
+      LOGGER.finest(Messages.get("LimitlessConnectionPlugin.limitlessRouterCacheEmpty"));
+      final boolean waitForRouterInfo = WAIT_F0R_ROUTER_INFO.getBoolean(props);
+      if (waitForRouterInfo) {
+        synchronouslyGetLimitlessRoutersWithRetry(conn, hostSpec.getPort(), props);
+      }
+    }
+
+    return conn;
+  }
+
   private void initLimitlessRouterMonitorService() {
     if (limitlessRouterService == null) {
       this.limitlessRouterService = this.limitlessRouterServiceSupplier.get();
     }
   }
 
-  private Connection retryConnectWithLeastLoadedRouters(final List<HostSpec> limitlessRouters,
-      final Properties props, final SQLException originalException) throws SQLException {
+  private Connection retryConnectWithLeastLoadedRouters(final List<HostSpec> limitlessRouters, final Properties props,
+      final Connection conn, final HostSpec hostSpec, final SQLException originalException) throws SQLException {
```

Comment:
Do we actually need to keep an original exception? What's the value it gives to a user to know that least-loaded-strategy has been caused by some node unavailability?

Reply:
Good point. Will remove it.